File: ADA/Laboratorio07/Ejercicio2.py
from math import floor, ceil
from ortools.linear_solver import pywraplp

def dfs(current, level, base, L, patterns):
    if level < len(base):
        for i in range(len(base)):
            current.append(base[i])

            partial = [*current]
            partial.sort()
            if sum(partial) <= L and L - sum(partial) < base[0] and not patterns.count(partial):
                patterns.append(partial)

            dfs(current, level+1, base, L, patterns)
            current.pop()

def generateMatrix(amountRows, amountColumns, initializer):
    mtx = []
    for i in range(amountRows):
        mtx.append([])
        for j in range(amountColumns):
            mtx[i].append(initializer)
    return mtx

def showMatrix(mtx):
    for i in range(len(mtx)):
        for j in range(len(mtx[i])):
            print(mtx[i][j], end=" ")
        print()

def solve(dp, q, patterns):
    solver = pywraplp.Solver.CreateSolver("GLOP")
    if not solver:
        return

    infinity = solver.infinity()
    vv = [solver.NumVar(0, infinity, f"x{j+1}") for j in range(len(dp[0]))]

    for i in range(len(dp)-1):
        solver.Add(sum([dp[i][j]*vv[j] for j in range(len(dp[i]))]) >= q[i])

    goal = solver.Objective()
    for j in range(len(dp[0])):
        goal.SetCoefficient(vv[j], dp[n][j])
    goal.SetMinimization()

    solver.Solve()

    solutions = [ceil(v.solution_value()) for v in vv]
    print("Frecuencia de los Patrones de Corte:")
    for j in range(len(solutions)):
        print(f"Patrón: {patterns[j]}")
        print(f"Frecuencia: {solutions[j]}")
        print()

    print()
    print(f"Desperdicio Mínimo: {ceil(solver.Objective().Value())}")

if __name__ == "__main__":
    L = 12
    n = 4
    d = [7, 8, 3, 5]
    q = [200, 85, 100, 150]

    print("Datos del problema:")
    print(f"L {L}")
    print(f"n: {n}")
    print(f"d (Dimensiones [m]): {d}")
    print(f"q (Demandas): {q}")
    print()
    print()

    for i in range(n):
        for j in range(n-i-1):
            if d[j] > d[j+1]:
                tmp = d[j]
                d[j] = d[j+1]
                d[j+1] = tmp

                tmp = q[j]
                q[j] = q[j+1]
                q[j+1] = tmp
    
    levels = floor(L/d[0])
    patterns = []
    dfs([], 0, d, L, patterns)

    dp = generateMatrix(n+1, len(patterns), 0)
    for j in range(len(patterns)):
        cc = 0
        for i in range(n):
            for k in range(len(patterns[j])):
                if d[i] == patterns[j][k]:
                    dp[i][j] = dp[i][j] + 1
    
    for j in range(len(patterns)):
        dp[n][j] = L - sum(patterns[j])
    
    solve(dp, q, patterns)

File: ADA/Laboratorio07/Ejercicio2.cpp
#include <iostream>
#include <cstddef>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;
using i64 = long long;

template <typename T>
void show(size_t n, T arr[]) {
    for(size_t i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << "\n";
}

template <typename T>
void show(vector<T> v) {
    for(T e: v) {
        cout << e << " ";
    }
    cout << "\n";
}

template <typename T>
void show(vector<vector<T>> mtx) {
    for(vector<T> v: mtx) {
        show<T>(v);
    }
}

template <typename T>
void show(set<vector<T>> mtx) {
    for(vector<T> v: mtx) {
        show<T>(v);
    }
}

void dfs(size_t cl, const size_t ml, vector<i64> &current, const size_t n, i64 base[], const i64 ll, i64 mm, vector<vector<i64>> &rpatterns) {
    if(cl <= ml) {
        for(size_t i = 0; i < n; i++) {
            current.push_back(base[i]);
            
            vector<i64> partial;
            for(i64 cc: current) {
                partial.push_back(cc);
            }
            i64 ss = 0;
            for(i64 pp: partial) {
                ss += pp;
            }
            if(ss <= ll && ll - ss < mm) {
                rpatterns.push_back(partial);
            }
            
            dfs(cl+1, ml, current, n, base, ll, mm, rpatterns);
            current.pop_back();            
        }
    }
}

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);

    const i64 ll = 12;
    const size_t n = 4;
    i64 d[n] = {5, 7, 8, 3};
    i64 q[n] = {150, 200, 85, 100};

    for(size_t i = 0; i < n; i++) {
        for(size_t j = 0; j < n - i - 1; j++) {
            if(d[j] > d[j + 1]) {
                i64 tmp = d[j];
                d[j] = d[j + 1];
                d[j + 1] = tmp;

                i64 tmp2 = q[j];
                q[j] = q[j + 1];
                q[j + 1] = tmp2;
            }
        }
    }

    const size_t levels = ll/d[0];
    vector<vector<i64>> rpatterns;
    vector<i64> current;
    dfs(1, levels, current, n, d, ll, d[0], rpatterns);
    set<vector<i64>> ppatterns;
    for(size_t i = 0; i < rpatterns.size(); i++) {
        sort(rpatterns[i].begin(), rpatterns[i].end());
        ppatterns.insert(rpatterns[i]);
    }
    vector<vector<i64>> patterns;
    for(vector<i64> v: ppatterns) {
        patterns.push_back(v);
    }

    i64 dp[n+1][patterns.size()]{};
    for(size_t j = 0; j < patterns.size(); j++) {
        for(size_t i = 0, counter = 0; i < n; i++) {
            while(patterns[j][counter] == d[i]) {
                dp[i][j]++;
                counter++;
            }
        }
    }

    for(size_t j = 0; j < patterns.size(); j++) {
        i64 ss = 0;
        for(i64 e: patterns[j]) {
            ss += e;
        }
        dp[n][j] = ll - ss;
    }

    for(size_t i = 0; i < n+1; i++) {
        for(size_t j = 0; j < patterns.size(); j++) {
            cout << dp[i][j] << " ";
        }
        cout << endl;
    }

    return 0;
}